src/Database/PostgreSQL/Transact/Cluster.hs

```diff
@@ -76,6 +78,10 @@ newtype CDBT (mode :: QueryMode) m a = CDBT {getDBT :: DBT m a}
         )
 
 
+-- The first argument of CDBT is nominal, because we don't want to allow coercions from CDBT ReadWrite to CDBT ReadOnly.
+type role CDBT nominal representational nominal
```

Comment:
Without this role definition, end-users are able to make dangerous coercions.  For instance, on `master`:

```haskell
> import Database.PostgreSQL.Transact.Cluster (CDBT, QueryMode(..))
> import Data.Coerce
> :set -XDataKinds
> :t coerce (undefined :: CDBT 'ReadWrite IO Int) :: CDBT 'ReadOnly IO Int
coerce (undefined :: CDBT 'ReadWrite IO Int) :: CDBT 'ReadOnly IO Int
```

But with this role definition, these coercions are no longer possible:

```haskell
> import Database.PostgreSQL.Transact.Cluster (CDBT, QueryMode(..))
> import Data.Coerce
> :set -XDataKinds
> :t coerce (undefined :: CDBT 'ReadWrite IO Int) :: CDBT 'ReadOnly IO Int
<interactive>:1:1: error:
    • Couldn't match type ‘'ReadWrite’ with ‘'ReadOnly’
        arising from a use of ‘coerce’
    • In the expression:
          coerce (undefined :: CDBT 'ReadWrite IO Int) ::
            CDBT 'ReadOnly IO Int
```

Reply:
@GambolingPangolin That was my first thought too, but when trying to set the last type parameter as `representational`, you get an error like the following:

```
src/Database/PostgreSQL/Transact/Cluster.hs:82:1: error:
    • Role mismatch on variable a:
        Annotation says representational but role nominal is required
    • while checking a role annotation for ‘CDBT’
   |
82 | type role CDBT nominal representational representational
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I think this is because of `DBT`'s roles:

```
> import Database.PostgreSQL.Transact
> :info DBT
type role DBT representational nominal
type DBT :: (* -> *) -> * -> *
newtype DBT m a
  = DBT {unDBT :: ReaderT Connection m a}
...
```

Similarly, `ReaderT` also has the last parameter as `nominal`:

```
> import Control.Monad.Trans.Reader 
*Database.PostgreSQL.Transact.Cluster.Connection Control.Monad.Trans.Reader > :info ReaderT
type role ReaderT representational representational nominal
type ReaderT :: * -> (* -> *) -> * -> *
newtype ReaderT r m a = ReaderT {runReaderT :: r -> m a}
...
```

I also checked `IdentityT`, and it is the same:

```
> import Control.Monad.Trans.Identity 
> :info IdentityT 
type role IdentityT representational nominal
type IdentityT :: forall {k}. (k -> *) -> k -> *
newtype IdentityT f a = IdentityT {runIdentityT :: f a}
...
```

It appears that the reason for this is that GHC requires all type variables that appear as an argument to another type variable, to have the role `nominal`.  This is explained in more detail in https://downloads.haskell.org/~ghc/9.4.3/docs/users_guide/exts/roles.html#role-inference.

Reply:
Very interesting!  Thanks for the explanation :^D